### PR TITLE
_cycleviz.py: visualize reference cycles holding cuda memory

### PR DESCRIPTION
**Pull request by zdevito**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102656

Reference cycles are freed by the cycle collector rather than being cleaned up
when the objects in the cycle first become unreachable. If a cycle points to a tensor,
the CUDA memory for that tensor will not be freed until garbage collection runs.
Accumulatin of CUDA allocations can lead to out of memory errors (OOMs), as well as
non-deterministic allocation behavior which is harder to debug.

This visualizer installs a garbage collection hook to look for cycles containing
CUDA tensors and saves a visualization of the garbage:

```
from torch.cuda._cycleviz import warn_tensor_cycles
warn_tensor_cycles()
# do some work that results in a cycle getting garbage collected
# ...
> WARNING:root:Reference cycle includes a CUDA Tensor see visualization of cycle /tmp/tmpeideu9gl.html
```

### CODE REVIEW

**Comment by zdevito** [comment]
@zdevito has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D46335835).

**Comment by zdevito** [comment]
@zdevito has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D46335835).

**Comment by zdevito** [comment]
@zdevito has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D46335835).

**Review by aaronenyeshi — APPROVED** [review]
This change looks great! We've verified that it works internally. Let's land this!

File: test/test_cuda.py [hunk 1]
```diff
@@ -5086,6 +5086,43 @@ def test_memory_profiler_viz(self):
...
+                return foo
+            create()
+            gc.collect()
+            # the callback has to run outside of the collect
+            # call so it doesn't actual fire until the next
+            # method call after a gc.collect
+            noop()
+            self.assertTrue(fired)
+        finally:
+            disarm()
```
Comment (aaronenyeshi) [inline test/test_cuda.py:5122]: Does having disarm allow us to toggle this on and off dynamically? By calling these two apis?
```
disarm = observe_tensor_cycles(observer)
disarm()
```

**Comment by zdevito** [comment]
@zdevito has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D46335835).

**Comment by huydhn** [comment]
@zdevito It looks this this change doesn't work on periodic Windows CUDA jobs in trunk https://github.com/pytorch/pytorch/actions/runs/5341701630/jobs/9683293600

```
2023-06-22T07:08:24.9352105Z __________________________ TestCudaComm.test_cycles ___________________________
2023-06-22T07:08:24.9352349Z Traceback (most recent call last):
2023-06-22T07:08:24.9352668Z   File "C:\actions-runner\_work\pytorch\pytorch\test\test_cuda.py", line 5090, in test_cycles
2023-06-22T07:08:24.9352978Z     disarm = observe_tensor_cycles(observer)
2023-06-22T07:08:24.9353533Z   File "C:\actions-runner\_work\pytorch\pytorch\build\win_tmp\build\torch\utils\viz\_cycles.py", line 423, in observe_tensor_cycles
2023-06-22T07:08:24.9354134Z     torch.cuda.memory._record_memory_history(max_entries=100000)
2023-06-22T07:08:24.9354775Z   File "C:\actions-runner\_work\pytorch\pytorch\build\win_tmp\build\torch\cuda\memory.py", line 676, in _record_memory_history
2023-06-22T07:08:24.9355188Z     return _record_memory_history_impl(enabled, *args, **kwargs)
2023-06-22T07:08:24.9355577Z   File "C:\actions-runner\_work\pytorch\pytorch\build\win_tmp\build\torch\cuda\memory.py", line 696, in _record_memory_history_impl
2023-06-22T07:08:24.9355973Z     _C._cuda_recordMemoryHistory(enabled_, record_context, record_context_cpp,
2023-06-22T07:08:24.9356302Z RuntimeError: record_context_cpp is not support on non-linux non-x86_64 platforms
```

Could you help take a look to see if it can be fixed?

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting your PR. But I think the change is failing on Windows CUDA https://github.com/pytorch/pytorch/actions/runs/5341701630/jobs/9683293600' -c nosignal

As this is a recent change, it would be easier to revert/rework than waiting for forward fix I think.

**Pull request closed without merging**